Astronomy/Astronomy/Helper Files/FetchPhotoOperation.m

```diff
@@ -0,0 +1,85 @@
+//
+//  FetchPhotoOperation.m
+//  Astronomy
+//
+//  Created by Karen Rodriguez on 5/19/20.
+//  Copyright © 2020 Wyatt Harrell. All rights reserved.
+//
+
+#import "FetchPhotoOperation.h"
+#import "Astronomy-Swift.h"
+
+@interface FetchPhotoOperation()
+
+@property BOOL executing;
+@property BOOL finished;
+//@property BOOL cancelled;
+@property BOOL asynchronous;
+
+@end
+
+@implementation FetchPhotoOperation
+
+@synthesize executing = _executing;
+@synthesize finished = _finished;
+//@synthesize cancelled = _cancelled;
+@synthesize asynchronous = _asynchronous;
+
+- (instancetype)initWithReference:(Photo *)photo {
+    self = [super init];
+    if (self) {
+        _photoReference = photo;
+        _imageData = nil;
+        _asynchronous = YES;
+    }
+    return self;
+}
+
+- (void)start {
+    // Tell observers we about to change this
+    [self willChangeValueForKey:@"isExecuting"];
+    self.executing = YES;
+    // Let observers know we changed this
+    [self didChangeValueForKey:@"isExecuting"];
+    
+    self.task = [NSURLSession.sharedSession dataTaskWithURL:self.photoReference.imgSrc completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
+        NSLog(@"Inside of FetchPhotoOperation with url: %@", self.photoReference.imgSrc);
+        if (error) {
+            NSLog(@"Error fetching image for photo: %@", self.photoReference);
+            [self willChangeValueForKey:@"isFinished"];
+            self.finished = YES;
+            [self didChangeValueForKey:@"isFinished"];
+            return;
+        }
+
+        self.imageData = data;
+
+        [self willChangeValueForKey:@"isFinished"];
+        self.finished = YES;
+        [self didChangeValueForKey:@"isFinished"];
+    }];
+    [self.task resume];
+}
+
+- (void)cancel {
+    [self willChangeValueForKey:@"isCancelled"];
+    [self.task cancel];
+    [super cancel];
+}
```

Comment:
I don't think this is necessary, use default implementation.

Reply:
@PaulSolt What exactly are you saying isn't necessary? I agree that sending the KVO notification is not needed, but this is needed as a hook to cancel the data task, as they are doing with `[self.task cancel]`. Also, you are responsible for setting isFinished according to the docs:

> In addition to simply exiting when an operation is cancelled, it is also important that you move a cancelled operation to the appropriate final state. Specifically, if you manage the values for the finished and executing properties yourself (perhaps because you are implementing a concurrent operation), you must update those properties accordingly. 

Not sure if they are doing this, but if not, the operation will be stuck in the queue when cancelled.

Reply:
My mistake. I was wrong when I first read this code.

This is good to override because we have a data task that is potentially active, so we can cancel it here. Overriding the setter for cancel shouldn't be necessary (I think that's what I was thinking about)

Thanks @swift-student